src/wlserver.cpp

```diff
@@ -486,6 +487,66 @@ static void create_gamescope_pipewire( void )
 }
 #endif
 
+static void gamescope_surface_tearing_set_presentation_hint( struct wl_client *client, struct wl_resource *resource, uint32_t hint )
+{
+	struct wlserver_surface *wlserver_surf = (struct wlserver_surface *)resource->data;
+
+	wlserver_surf->presentation_hint = hint;
+	wl_log.errorf( "WE GOT A TEARING HINT! %u", hint );
+}
+
+static void gamescope_surface_tearing_destroy( struct wl_client *client, struct wl_resource *resource )
+{
+	wl_resource_destroy( resource );
+}
+
+static const struct zwp_surface_tearing_control_v1_interface surface_tearing_control_impl {
+	.set_presentation_hint = gamescope_surface_tearing_set_presentation_hint,
+	.destroy = gamescope_surface_tearing_destroy,
+};
+
+static void gamescope_tearing_handle_destroy( struct wl_client *client, struct wl_resource *resource )
+{
+	wl_resource_destroy( resource );
+}
+
+static void gamescope_tearing_get_tearing_control( struct wl_client *client, struct wl_resource *resource, uint32_t id, struct wl_resource *surface_resource )
+{
+	struct wlr_surface *surface = wlr_surface_from_resource( surface_resource );
+	struct wlserver_surface *wlserver_surf = (struct wlserver_surface *)surface->data;
+
+	if ( !wlserver_surf )
+	{
+		wl_log.errorf( "Failed to get tearing control as surface was not associated with a wlserver_surface." );
+		return;
+	}
+	
+	struct wl_resource *surface_tearing_control_resource
+		= wl_resource_create( client, &zwp_surface_tearing_control_v1_interface, 1, id );
```

Comment:
For future proofing, instead of hardcoding 1, we should use `wl_resource_get_version(resource)` here, to inherit the version of the manager object.